File: src-tauri/src/cache.rs
/// LRU cache with TTL for parsed session data.

use std::collections::HashMap;
use std::num::NonZeroUsize;
use std::time::{Duration, Instant};

use lru::LruCache;

use crate::parsing::session_parser::SessionFileMetadata;
use crate::types::domain::ParsedSession;

/// Hard memory budget for the whole session cache (sprint 63).
/// When `total_byte_estimate` exceeds this, oldest entries are evicted from
/// the LRU tail until the cache is back under budget. The bytes value is a
/// coarse heuristic (≈2 KB / message + parsed-session overhead); under-
/// estimation is preferred over double-counting so the budget acts as a
/// ceiling, not a floor.
pub const MAX_CACHE_BYTES: usize = 200 * 1024 * 1024;
const BYTES_PER_MESSAGE_ESTIMATE: usize = 2048;
const BASE_SESSION_BYTES: usize = 4096;

fn estimate_session_bytes(value: &ParsedSession) -> usize {
    BASE_SESSION_BYTES + value.messages.len() * BYTES_PER_MESSAGE_ESTIMATE
}

/// Cached entry with expiration tracking + coarse byte-budget hint.
struct CacheEntry {
    value: ParsedSession,
    inserted_at: Instant,
    byte_estimate: usize,
}

/// Tracks incremental parsing state for a session file.
#[derive(Debug, Clone)]
pub struct IncrementalState {
    /// Byte offset of the last successfully parsed position.
    pub byte_offset: u64,
    /// Accumulated metadata from all lines parsed so far.
    pub metadata: SessionFileMetadata,
}

pub struct SessionCache {
    inner: LruCache<String, CacheEntry>,
    ttl: Duration,
    /// Tracks incremental parsing state per session (keyed by cache key).
    incremental: HashMap<String, IncrementalState>,
    /// Hit/miss/evict counters (sprint 46).
    pub hits: u64,
    pub misses: u64,
    pub evicts: u64,
    /// Sprint 63: byte-budget eviction counter (separate from LRU evicts).
    pub budget_evicts: u64,
    /// Sprint 63: total estimated bytes resident in the cache.
    total_byte_estimate: usize,
    /// Sprint 63: hard memory cap; insertions evict from the LRU tail.
    max_bytes: usize,
}

impl SessionCache {
    pub fn new(capacity: usize, ttl: Duration) -> Self {
        Self {
            inner: LruCache::new(NonZeroUsize::new(capacity).unwrap_or(NonZeroUsize::new(50).unwrap())),
            ttl,
            incremental: HashMap::new(),
            hits: 0,
            misses: 0,
            evicts: 0,
            budget_evicts: 0,
            total_byte_estimate: 0,
            max_bytes: MAX_CACHE_BYTES,
        }
    }

    pub fn total_byte_estimate(&self) -> usize {
        self.total_byte_estimate
    }

    pub fn max_bytes(&self) -> usize {
        self.max_bytes
    }

    pub fn set_max_bytes(&mut self, bytes: usize) {
        self.max_bytes = bytes.max(1024 * 1024);
        self.enforce_byte_budget();
    }

    fn enforce_byte_budget(&mut self) {
        while self.total_byte_estimate > self.max_bytes && self.inner.len() > 0 {
            if let Some((_, evicted)) = self.inner.pop_lru() {
                self.total_byte_estimate = self
                    .total_byte_estimate
                    .saturating_sub(evicted.byte_estimate);
                self.budget_evicts += 1;
            } else {
                break;
            }
        }
    }

    pub fn get(&mut self, key: &str) -> Option<&ParsedSession> {
        let expired = match self.inner.get(key) {
            Some(e) => e.inserted_at.elapsed() > self.ttl,
            None => {
                self.misses += 1;
                return None;
            }
        };
        if expired {
            if let Some(entry) = self.inner.pop(key) {
                self.total_byte_estimate = self
                    .total_byte_estimate
                    .saturating_sub(entry.byte_estimate);
            }
            self.evicts += 1;
            self.misses += 1;
            return None;
        }
        self.hits += 1;
        self.inner.get(key).map(|e| &e.value)
    }

    pub fn insert(&mut self, key: String, value: ParsedSession) {
        let was_full = self.inner.len() == self.inner.cap().get();
        let byte_estimate = estimate_session_bytes(&value);
        let prior = self.inner.pop(&key);
        if let Some(p) = &prior {
            self.total_byte_estimate = self.total_byte_estimate.saturating_sub(p.byte_estimate);
        }
        self.total_byte_estimate += byte_estimate;
        self.inner.put(
            key,
            CacheEntry {
                value,
                inserted_at: Instant::now(),
                byte_estimate,
            },
        );
        if was_full && prior.is_none() {
            self.evicts += 1;
        }
        self.enforce_byte_budget();
    }

    /// Sprint 46: hot-resize the cache. New capacity ≥ 1; entries beyond
    /// the new bound are evicted from the LRU tail.
    pub fn set_capacity(&mut self, capacity: usize) {
        let cap = NonZeroUsize::new(capacity.max(1)).unwrap();
        let prior = self.inner.len();
        self.inner.resize(cap);
        let new_len = self.inner.len();
        if prior > new_len {
            self.evicts += (prior - new_len) as u64;
        }
    }

    pub fn capacity(&self) -> usize {
        self.inner.cap().get()
    }

    pub fn len(&self) -> usize {
        self.inner.len()
    }

    pub fn clear(&mut self) {
        self.evicts += self.inner.len() as u64;
        self.inner.clear();
        self.incremental.clear();
        self.total_byte_estimate = 0;
    }

    pub fn get_incremental(&self, key: &str) -> Option<&IncrementalState> {
        self.incremental.get(key)
    }

    pub fn set_incremental(&mut self, key: String, state: IncrementalState) {
        self.incremental.insert(key, state);
    }

    pub fn remove_incremental(&mut self, key: &str) {
        self.incremental.remove(key);
    }

    // Mirrors Go `SessionCache.InvalidateProject`: evicts every cached session
    // whose key is under the encoded project dir. The W13 maintenance trash path
    // is the first production caller (`evict_trashed_projects`).
    pub fn invalidate_project(&mut self, project_id: &str) {
        let prefix = format!("{project_id}/");
        let keys_to_remove: Vec<String> = self
            .inner
            .iter()
            .filter(|(k, _)| k.starts_with(&prefix))
            .map(|(k, _)| k.clone())
            .collect();
        for key in &keys_to_remove {
            if let Some(entry) = self.inner.pop(key) {
                self.total_byte_estimate = self
                    .total_byte_estimate
                    .saturating_sub(entry.byte_estimate);
            }
            self.incremental.remove(key.as_str());
        }
    }
}

#[cfg(test)]
impl SessionCache {
    pub fn invalidate(&mut self, key: &str) {
        if let Some(entry) = self.inner.pop(key) {
            self.total_byte_estimate = self
                .total_byte_estimate
                .saturating_sub(entry.byte_estimate);
        }
        self.incremental.remove(key);
    }
}

impl Default for SessionCache {
    fn default() -> Self {
        Self::new(50, Duration::from_secs(600))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::domain::{MessagesByType, SessionMetrics};

    fn make_session() -> ParsedSession {
        ParsedSession {
            messages: vec![],
            metrics: SessionMetrics::default(),
            task_calls: vec![],
            by_type: MessagesByType {
                user: vec![],
                real_user: vec![],
                internal_user: vec![],
                assistant: vec![],
                system: vec![],
                other: vec![],
            },
            sidechain_messages: vec![],
            main_messages: vec![],
            custom_title: None,
            agent_name: None,
        }
    }

    #[test]
    fn test_insert_and_get() {
        let mut cache = SessionCache::default();
        cache.insert("proj/sess1".to_string(), make_session());
        assert!(cache.get("proj/sess1").is_some());
        assert!(cache.get("proj/sess2").is_none());
    }

    #[test]
    fn test_invalidate() {
        let mut cache = SessionCache::default();
        cache.insert("proj/sess1".to_string(), make_session());
        cache.invalidate("proj/sess1");
        assert!(cache.get("proj/sess1").is_none());
    }

    #[test]
    fn test_invalidate_project() {
        let mut cache = SessionCache::default();
        cache.insert("proj1/sess1".to_string(), make_session());
        cache.insert("proj1/sess2".to_string(), make_session());
        cache.insert("proj2/sess1".to_string(), make_session());
        cache.invalidate_project("proj1");
        assert!(cache.get("proj1/sess1").is_none());
        assert!(cache.get("proj1/sess2").is_none());
        assert!(cache.get("proj2/sess1").is_some());
    }

    #[test]
    fn test_ttl_expiration() {
        let mut cache = SessionCache::new(50, Duration::from_millis(1));
        cache.insert("proj/sess1".to_string(), make_session());
        std::thread::sleep(Duration::from_millis(10));
        assert!(cache.get("proj/sess1").is_none());
    }

    // =========================================================================
    // Incremental state tracking
    // =========================================================================

    #[test]
    fn test_set_and_get_incremental() {
        let mut cache = SessionCache::default();
        let state = IncrementalState {
            byte_offset: 1024,
            metadata: SessionFileMetadata {
                custom_title: Some("My Session".to_string()),
                agent_name: None,
            },
        };
        cache.set_incremental("proj/sess1".to_string(), state);
        let retrieved = cache.get_incremental("proj/sess1").unwrap();
        assert_eq!(retrieved.byte_offset, 1024);
        assert_eq!(retrieved.metadata.custom_title.as_deref(), Some("My Session"));
    }

    #[test]
    fn test_get_incremental_missing() {
        let cache = SessionCache::default();
        assert!(cache.get_incremental("nonexistent").is_none());
    }

    #[test]
    fn test_remove_incremental() {
        let mut cache = SessionCache::default();
        cache.set_incremental(
            "proj/sess1".to_string(),
            IncrementalState {
                byte_offset: 512,
                metadata: SessionFileMetadata::default(),
            },
        );
        cache.remove_incremental("proj/sess1");
        assert!(cache.get_incremental("proj/sess1").is_none());
    }

    #[test]
    fn test_update_incremental_offset() {
        let mut cache = SessionCache::default();
        cache.set_incremental(
            "proj/sess1".to_string(),
            IncrementalState {
                byte_offset: 100,
                metadata: SessionFileMetadata::default(),
            },
        );
        cache.set_incremental(
            "proj/sess1".to_string(),
            IncrementalState {
                byte_offset: 500,
                metadata: SessionFileMetadata {
                    custom_title: Some("Updated".to_string()),
                    agent_name: None,
                },
            },
        );
        let state = cache.get_incremental("proj/sess1").unwrap();
        assert_eq!(state.byte_offset, 500);
        assert_eq!(state.metadata.custom_title.as_deref(), Some("Updated"));
    }

    #[test]
    fn test_invalidate_clears_incremental() {
        let mut cache = SessionCache::default();
        cache.insert("proj/sess1".to_string(), make_session());
        cache.set_incremental(
            "proj/sess1".to_string(),
            IncrementalState {
                byte_offset: 256,
                metadata: SessionFileMetadata::default(),
            },
        );
        cache.invalidate("proj/sess1");
        assert!(cache.get("proj/sess1").is_none());
        assert!(cache.get_incremental("proj/sess1").is_none());
    }

    fn session_with_messages(n: usize) -> ParsedSession {
        let mut s = make_session();
        s.messages = (0..n)
            .map(|i| crate::types::messages::ParsedMessage {
                uuid: format!("u{i}"),
                parent_uuid: None,
                message_type: "user".to_string(),
                timestamp: "2024-01-01T00:00:00Z".to_string(),
                role: Some("user".to_string()),
                content: crate::types::messages::ParsedMessageContent::Text("hi".to_string()),
                usage: None,
                model: None,
                cwd: None,
                git_branch: None,
                agent_id: None,
                is_sidechain: false,
                is_meta: false,
                user_type: None,
                tool_calls: vec![],
                tool_results: vec![],
                source_tool_use_id: None,
                source_tool_assistant_uuid: None,
                tool_use_result: None,
                is_compact_summary: None,
                request_id: None,
                subtype: None,
                event_data: None,
            })
            .collect();
        s
    }

    #[test]
    fn test_byte_estimate_increments_on_insert() {
        let mut cache = SessionCache::default();
        let before = cache.total_byte_estimate();
        cache.insert("proj/s1".to_string(), session_with_messages(10));
        assert!(cache.total_byte_estimate() > before);
    }

    #[test]
    fn test_byte_budget_evicts_lru_tail() {
        let mut cache = SessionCache::default();
        // 1 MB cap: forces eviction quickly.
        cache.set_max_bytes(1024 * 1024);

        // Each session ≈ 100 messages * 2KB = 200KB + base. Insert 10 to overflow.
        for i in 0..10 {
            cache.insert(format!("p/s{i}"), session_with_messages(100));
        }
        assert!(
            cache.total_byte_estimate() <= cache.max_bytes(),
            "total bytes ({}) must stay under cap ({})",
            cache.total_byte_estimate(),
            cache.max_bytes()
        );
        assert!(
            cache.budget_evicts > 0,
            "budget_evicts counter must record evictions"
        );
        // Earliest inserts should have been evicted (LRU tail).
        assert!(cache.get("p/s0").is_none());
    }

    #[test]
    fn test_set_max_bytes_re_enforces_immediately() {
        let mut cache = SessionCache::default();
        // 10 sessions * 500 msgs * 2KB ≈ 10 MB total to clearly exceed clamp floor.
        for i in 0..10 {
            cache.insert(format!("p/s{i}"), session_with_messages(500));
        }
        let before_total = cache.total_byte_estimate();
        let before_evicts = cache.budget_evicts;
        cache.set_max_bytes(1024 * 1024); // 1 MB cap (the minimum floor)
        assert!(cache.budget_evicts > before_evicts);
        assert!(cache.total_byte_estimate() < before_total);
        assert!(cache.total_byte_estimate() <= cache.max_bytes());
    }

    #[test]
    fn test_invalidate_project_clears_incremental() {
        let mut cache = SessionCache::default();
        cache.insert("proj1/sess1".to_string(), make_session());
        cache.set_incremental(
            "proj1/sess1".to_string(),
            IncrementalState {
                byte_offset: 100,
                metadata: SessionFileMetadata::default(),
            },
        );
        cache.insert("proj2/sess1".to_string(), make_session());
        cache.set_incremental(
            "proj2/sess1".to_string(),
            IncrementalState {
                byte_offset: 200,
                metadata: SessionFileMetadata::default(),
            },
        );
        cache.invalidate_project("proj1");
        assert!(cache.get_incremental("proj1/sess1").is_none());
        assert!(cache.get_incremental("proj2/sess1").is_some());
    }
}

File: internal/cache/cache.go
// Package cache: an LRU cache with TTL + a coarse
// byte-budget ceiling for parsed session data, plus per-session incremental
// parse state. The cache owns its mutex internally; callers never lock it
// (arch C1). Layering: imports only domain + parsing, never pipeline/analysis/
// any service (arch H1) — enforced by import_test.go.
package cache

import (
	"strings"
	"sync"
	"time"

	lru "github.com/hashicorp/golang-lru/v2/simplelru"

	"claude-devtools/internal/domain"
	"claude-devtools/internal/parsing"
)

// MaxCacheBytes is the hard memory budget for the whole session cache
// (cache.rs:18). When the byte estimate exceeds it, oldest entries are evicted
// from the LRU tail until back under budget.
const MaxCacheBytes = 200 * 1024 * 1024

const (
	bytesPerMessageEstimate = 2048
	baseSessionBytes        = 4096
	defaultCapacity         = 50
	defaultTTL              = 600 * time.Second
	minMaxBytes             = 1024 * 1024
)

func estimateSessionBytes(v domain.ParsedSession) int {
	return baseSessionBytes + len(v.Messages)*bytesPerMessageEstimate
}

func saturatingSub(a, b int) int {
	if b > a {
		return 0
	}
	return a - b
}

type cacheEntry struct {
	value        domain.ParsedSession
	insertedAt   time.Time
	byteEstimate int
}

// IncrementalState tracks incremental parsing state for a session file
// (cache.rs:35).
type IncrementalState struct {
	// ByteOffset is the byte offset of the last successfully parsed position.
	ByteOffset uint64
	// Metadata is the accumulated metadata from all lines parsed so far.
	Metadata parsing.SessionFileMetadata
}

// Stats is the snapshot read by TimingService.get_cache_stats (timing.rs:149).
type Stats struct {
	Capacity int     `json:"capacity"`
	Size     int     `json:"size"`
	Hits     uint64  `json:"hits"`
	Misses   uint64  `json:"misses"`
	Evicts   uint64  `json:"evicts"`
	HitRate  float64 `json:"hitRate"`
}

// SessionCache is goroutine-safe; every method takes the internal mutex.
type SessionCache struct {
	mu          sync.Mutex
	inner       *lru.LRU[string, *cacheEntry]
	capacity    int // simplelru has no Cap() accessor; track it ourselves
	ttl         time.Duration
	incremental map[string]IncrementalState

	hits         uint64
	misses       uint64
	evicts       uint64
	budgetEvicts uint64

	totalByteEstimate int
	maxBytes          int
}

// New mirrors SessionCache::new (cache.rs:60); capacity < 1 falls back to 50.
func New(capacity int, ttl time.Duration) *SessionCache {
	if capacity < 1 {
		capacity = defaultCapacity
	}
	// nil onEvict: byte accounting is done manually (as in Rust) so set_capacity's
	// auto-eviction does NOT decrement the estimate — matching cache.rs exactly.
	inner, _ := lru.NewLRU[string, *cacheEntry](capacity, nil)
	return &SessionCache{
		inner:       inner,
		capacity:    capacity,
		ttl:         ttl,
		incremental: make(map[string]IncrementalState),
		maxBytes:    MaxCacheBytes,
	}
}

// Default mirrors SessionCache::default (cache.rs:216): 50 entries, 10 min TTL.
func Default() *SessionCache { return New(defaultCapacity, defaultTTL) }

func (c *SessionCache) TotalByteEstimate() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.totalByteEstimate
}

func (c *SessionCache) MaxBytes() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.maxBytes
}

func (c *SessionCache) BudgetEvicts() uint64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.budgetEvicts
}

// SetMaxBytes clamps to a 1 MB floor then re-enforces immediately (cache.rs:82).
func (c *SessionCache) SetMaxBytes(bytes int) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if bytes < minMaxBytes {
		bytes = minMaxBytes
	}
	c.maxBytes = bytes
	c.enforceByteBudget()
}

// enforceByteBudget evicts from the LRU tail while over budget. Caller holds mu.
func (c *SessionCache) enforceByteBudget() {
	for c.totalByteEstimate > c.maxBytes && c.inner.Len() > 0 {
		_, evicted, ok := c.inner.RemoveOldest()
		if !ok {
			break
		}
		c.totalByteEstimate = saturatingSub(c.totalByteEstimate, evicted.byteEstimate)
		c.budgetEvicts++
	}
}

// Get returns the cached session and true on a live hit; expired entries are
// evicted and counted as a miss (cache.rs:100).
func (c *SessionCache) Get(key string) (domain.ParsedSession, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	entry, ok := c.inner.Get(key)
	if !ok {
		c.misses++
		return domain.ParsedSession{}, false
	}
	if time.Since(entry.insertedAt) > c.ttl {
		c.inner.Remove(key)
		c.totalByteEstimate = saturatingSub(c.totalByteEstimate, entry.byteEstimate)
		c.evicts++
		c.misses++
		return domain.ParsedSession{}, false
	}
	c.hits++
	return entry.value, true
}

// Insert stores a session, replacing any prior entry for the key (cache.rs:122).
func (c *SessionCache) Insert(key string, value domain.ParsedSession) {
	c.mu.Lock()
	defer c.mu.Unlock()
	wasFull := c.inner.Len() == c.capacity
	byteEstimate := estimateSessionBytes(value)
	prior, hadPrior := c.inner.Peek(key)
	c.inner.Remove(key) // pop the prior entry, mirroring lru::pop
	if hadPrior {
		c.totalByteEstimate = saturatingSub(c.totalByteEstimate, prior.byteEstimate)
	}
	c.totalByteEstimate += byteEstimate
	c.inner.Add(key, &cacheEntry{value: value, insertedAt: time.Now(), byteEstimate: byteEstimate})
	if wasFull && !hadPrior {
		c.evicts++
	}
	c.enforceByteBudget()
}

// SetCapacity hot-resizes; entries beyond the new bound are evicted from the
// LRU tail (cache.rs:146). Byte estimate is intentionally NOT adjusted here, to
// match the Rust behavior exactly.
func (c *SessionCache) SetCapacity(capacity int) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if capacity < 1 {
		capacity = 1
	}
	prior := c.inner.Len()
	c.inner.Resize(capacity)
	c.capacity = capacity
	newLen := c.inner.Len()
	if prior > newLen {
		c.evicts += uint64(prior - newLen)
	}
}

func (c *SessionCache) Capacity() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.capacity
}

func (c *SessionCache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.inner.Len()
}

// Clear empties the cache + incremental state (cache.rs:164).
func (c *SessionCache) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.evicts += uint64(c.inner.Len())
	c.inner.Purge()
	c.incremental = make(map[string]IncrementalState)
	c.totalByteEstimate = 0
}

// GetStats returns the live hit-rate snapshot (timing.rs:149).
func (c *SessionCache) GetStats() Stats {
	c.mu.Lock()
	defer c.mu.Unlock()
	total := c.hits + c.misses
	hitRate := 0.0
	if total != 0 {
		hitRate = float64(c.hits) / float64(total)
	}
	return Stats{
		Capacity: c.capacity,
		Size:     c.inner.Len(),
		Hits:     c.hits,
		Misses:   c.misses,
		Evicts:   c.evicts,
		HitRate:  hitRate,
	}
}

func (c *SessionCache) GetIncremental(key string) (IncrementalState, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	s, ok := c.incremental[key]
	return s, ok
}

func (c *SessionCache) SetIncremental(key string, state IncrementalState) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.incremental[key] = state
}

func (c *SessionCache) RemoveIncremental(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.incremental, key)
}

// Invalidate removes a single entry + its incremental state (cache.rs:187).
func (c *SessionCache) Invalidate(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if entry, ok := c.inner.Peek(key); ok {
		c.inner.Remove(key)
		c.totalByteEstimate = saturatingSub(c.totalByteEstimate, entry.byteEstimate)
	}
	delete(c.incremental, key)
}

// InvalidateProject removes every entry whose key is prefixed "<projectID>/"
// plus their incremental state (cache.rs:196).
func (c *SessionCache) InvalidateProject(projectID string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	prefix := projectID + "/"
	var toRemove []string
	for _, k := range c.inner.Keys() {
		if strings.HasPrefix(k, prefix) {
			toRemove = append(toRemove, k)
		}
	}
	for _, k := range toRemove {
		if entry, ok := c.inner.Peek(k); ok {
			c.inner.Remove(k)
			c.totalByteEstimate = saturatingSub(c.totalByteEstimate, entry.byteEstimate)
		}
		delete(c.incremental, k)
	}
}
